core/consensus/grandpa/impl/environment_impl.cpp

```diff
@@ -0,0 +1,156 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "consensus/grandpa/impl/environment_impl.hpp"
+
+#include <utility>
+
+#include <boost/optional/optional_io.hpp>
+#include "scale/scale.hpp"
+
+OUTCOME_CPP_DEFINE_CATEGORY(kagome::consensus::grandpa,
+                            EnvironmentImpl::Error,
+                            e) {
+  using E = kagome::consensus::grandpa::EnvironmentImpl::Error;
+  switch (e) {
+    case E::BLOCK_AFTER_LIMIT:
+      return "target block is after limit";
+  }
+  return "Unknown error";
+}
+
+namespace kagome::consensus::grandpa {
+
+  using primitives::BlockHash;
+  using primitives::BlockNumber;
+
+  EnvironmentImpl::EnvironmentImpl(
+      std::shared_ptr<blockchain::BlockTree> block_tree,
+      std::shared_ptr<blockchain::BlockHeaderRepository> header_repository,
+      std::shared_ptr<Gossiper> gossiper)
+      : block_tree_{std::move(block_tree)},
+        header_repository_{std::move(header_repository)},
+        gossiper_{std::move(gossiper)},
+        logger_{common::createLogger("Grandpa environment:")} {
+    BOOST_ASSERT(block_tree_ != nullptr);
+    BOOST_ASSERT(header_repository_ != nullptr);
+    BOOST_ASSERT(gossiper_ != nullptr);
+  }
+
+  outcome::result<std::vector<BlockHash>> EnvironmentImpl::getAncestry(
+      const BlockHash &base, const BlockHash &block) const {
+    OUTCOME_TRY(chain, block_tree_->getChainByBlocks(base, block));
+    std::vector<BlockHash> result_chain(chain.size() - 2);
+    std::copy(chain.rbegin() + 1, chain.rend() - 1, result_chain.begin());
+    return result_chain;
+  }
+
+  outcome::result<BlockInfo> EnvironmentImpl::bestChainContaining(
+      const BlockHash &base) const {
+    boost::optional<uint64_t> limit =
```

Comment:
Btw did you by chance have any idea about this?

Reply:
no(